---
title: "Vitiligo age-of-onset and PRS Association"
author: "Genevieve HL Roberts"
date: "2024-8-1"
output:
  pdf_document:
    keep_tex: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, results = 'asis', message = FALSE, warning = FALSE)
```

```{r packages, include=FALSE}
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)
library(tibble)
library(ggplot2)
library(broom)
library(here)
library(mixtools)
library(gridExtra)
library(lubridate)
library(pandoc)
library(pander)
library(car)
library(data.table)
library(patchwork)

source(here("functions.R"))
```

# Introduction

In this study, we aim to investigate the association between a vitiligo 
Polygenic Risk Score (PRS) excluding MHC Class II region SNPs and the 
age-of-onset. Our hypothesis posits a correlation wherein a higher PRS 
(excluding MHC Class II) is linked to an earlier age-of-onset, 
indicating a potential association with elevated genetic risk.

Moreover, we extend our hypothesis to carriers of the extreme risk/early age 
of onset MHC Class II haplotype. In this subgroup, we anticipate that the 
impact of a high PRS without MHC Class II may be diminished due to the 
substantial genetic risk conferred by the haplotype (haplotype OR=8.1). 
We believe there may be an interaction effect, suggesting that the slope 
of the regression line for age-of-onset ~ PRS is smaller in individuals 
carrying the high-risk haplotype compared to non-carriers.


```{r readin}
# Read genotypes data
risk_genos <- read.table(str_c(here(), "/sensitive_data/GWAS123_forRiskScore_inR.csv"), sep = ",", header = TRUE) %>%
  mutate(vitiligo = ifelse(PHENOTYPE==2, 1, 0)) %>%
  mutate(sex = ifelse(SEX==2, 1, 0))  %>%
  mutate(IID = toupper(IID)) %>%
  mutate(eAOO_hap_carrier =
           case_when(
             RS145954018 >= 1 & RS114448410 >=1 ~ 1L,
             RS145954018 >= 1 ~ NA_integer_,
             TRUE ~ 0))

# Read in PCs
pcs <- read.table(here("sensitive_data/GWAS123_DanielIDs_Phenos_PCs.txt"), header=TRUE) %>%
  select(IID, starts_with("g")) %>%
  mutate(IID = toupper(IID))

kCovariates <- pcs %>%
  select(starts_with("g")) %>%
  names()
kCovariates <- c(kCovariates, "sex")

#read in AOO
aoo <- read.table(here("sensitive_data/GWAS123G3RepageonsetonlyFinal_from_Ying_AOO.csv"),
                  header=TRUE, sep=",") %>%
  mutate(IID = toupper(newID)) %>%
  select(IID, VITageonset)

# Read in year of onset data
yoo <- read.table(here("sensitive_data/AgeonsetDurationExpansion.csv"),
                  header=TRUE, sep=",") %>%
  mutate(IID = toupper(newID)) %>%
  mutate(birth_year = as.numeric(str_extract(DOB, "\\d+$"))) %>%
  mutate(birth_year =
           ifelse(birth_year >= 5 & birth_year <= 99,
                  birth_year + 1900, birth_year + 2000)) %>%
  mutate(year_of_onset = birth_year + VITageonset) %>%
  select(IID, year_of_onset)

# Read risk score weights
weights <- read.table(here("Risk_Score_Weights_no_TagSNPs.csv"),
                      sep = ",", header = TRUE) %>%
  #removes the MHC Class II SNPs, which has special coding
  filter(!SNP %in% c("RS114448410", "RS145954018:RS114448410"))
```

```{r recode alleles}
# Recode SNPs to make sure everything is w/r/t risk allele
snps_to_recode <- weights %>%
  filter(Recode. == "Yes") %>%
  pull(SNP)

risk_genos_recoded <- recode_snps(df=risk_genos,
                                  snps_to_recode=snps_to_recode)
```

```{r compute PRS}
# compute the main PRS (does not include MHC Class II)
main_prs_no_mhcII <- calculate_main_prs(risk_genos = risk_genos_recoded,
                           weights = weights,
                           iid_col = "IID",
                           weights_col = "Weights_risk_direction_only") %>%
  rename(no_mhc_classII_prs = score)

#compute MHC only PRS
mhcII_only_prs <- calculate_class2_risk_score(risk_genos = risk_genos_recoded)

#join with the main dataframe and with AOO data
analysis_df <- risk_genos_recoded %>%
  left_join(main_prs_no_mhcII) %>%
  left_join(mhcII_only_prs) %>%
  left_join(pcs) %>%
  left_join(aoo) %>%
  left_join(yoo) %>%
  mutate(CONFIRMED_prs = no_mhc_classII_prs + mhc_class2_only_prs)

#select the columns that are prs
prs_cols <- analysis_df %>%
  select(ends_with("prs")) %>%
  names()

#scale the PRS with respect to the mean and sd in controls
analysis_df <- scale_prs(data = analysis_df,
                         pheno_col = "vitiligo",
                         pheno_cond = 0,
                         scale_columns = prs_cols) %>%
  filter(!is.na(CONFIRMED_prs))

#make a case only analysis df with non-missing age-of-onset
case_only_analysis_df <- analysis_df %>%
  filter(!is.na(VITageonset))
```

\newpage
# Results

## Categorizing age-of-onset into Early- and Late- Onset Groups
I no longer have the AOO classifications that we created previously, so I 
have re-fit the finite mixture model to categorize GWAS123 cases as either
early onset or late onset.  I have used the same methods here.

In the eAOO paper it says,
"After subgroup assignment and control matching, 
the combined early-onset subgroup contained 704 cases and 9,031 controls 
and the combined late-onset subgroup contained 1,467 cases and 19,156 controls."

```{r fmm to get aoo groups}
# fit a finite mixture model with 2 groups to get early and late
# onset group designations
set.seed(123)
fmm_mat <- as.matrix(case_only_analysis_df$VITageonset)
rownames(fmm_mat) <- case_only_analysis_df$IID

fit_fmm <- normalmixEM(fmm_mat, k = 2)
eaoo_posterior <- bind_cols(case_only_analysis_df$IID,
                            fit_fmm$posterior[,1],
                            fit_fmm$posterior[,2]) 
colnames(eaoo_posterior) <- c("IID",
                              "eAOO_posterior_probability",
                              "lAOO_posterior_probability")

#categorize as "eAOO" if posterior probability of eAOO group is >=80%
eaoo_posterior <- eaoo_posterior %>%
  mutate(AOO_category = 
           case_when(
             eAOO_posterior_probability >= 0.8 ~ "early_onset",
             lAOO_posterior_probability >= 0.8 ~ "late_onset",
             TRUE ~ NA_character_
           ))


analysis_df <- analysis_df %>%
  left_join(eaoo_posterior) %>%
  mutate(early_onset_vitiligo =
           case_when(
             AOO_category == "early_onset" & vitiligo == 1 ~ 1L,
             vitiligo == 0 ~ 0L,
             TRUE ~ NA_integer_
           ),
         late_onset_vitiligo =
           case_when(
             AOO_category == "late_onset" & vitiligo == 1 ~ 1L,
             vitiligo == 0 ~ 0L,
             TRUE ~ NA_integer_
           ),
         eA00_rs145954018 = RS145954018,
         generic_rs9271597 = RS114448410
         )

case_only_analysis_df <- analysis_df %>%
  filter(!is.na(VITageonset))

summarize_aoo_cat <- case_only_analysis_df %>%
  group_by(AOO_category) %>%
  summarize(n = n(),
            `Actual Median` = median(VITageonset),
            `Actual Mean` = mean(VITageonset),
            `Actual Minimum` = min(VITageonset),
            `Actual Maximum` = max(VITageonset),
            `Actual SD` = sd(VITageonset)
  ) %>%
  mutate(`FMM Mean` = ifelse(AOO_category == "early_onset", fit_fmm$mu[1],
                           ifelse(AOO_category == "late_onset", fit_fmm$mu[2], NA)),
         `FMM SD` = ifelse(AOO_category == "early_onset", fit_fmm$sigma[1],
                           ifelse(AOO_category == "late_onset", fit_fmm$sigma[2], NA))) %>%
  transmute(
    `Age of Onset Category` = AOO_category,
    n,
    `FMM Mean`,
    `Actual Mean`,
    `FMM SD`,
    `Actual SD`,
    `Actual Minimum`,
    `Actual Maximum`
  )

pandoc.table(summarize_aoo_cat,
             round = 2,
             caption="Early and Late Onset Subgroup Descriptive Statistics")
```

As shown in the table above, you can see that we achieve similar sample sizes
for early and late onset (719 early here vs. 704 in our earlier analysis
and 1,431 late here vs. 1,467 in our earlier analysis).The FMM-estimated 
distribution means are also identical, at 10.3 here
versus 10.3 for GWAS123 + Rep in our earlier analysis and 34.0
here versus 34.0 for GWAS123 + Rep in our earlier analysis.

\newpage
```{r}
dist_params  <- summarize_aoo_cat %>%
  filter(!is.na(`FMM Mean`) & !is.na(`FMM SD`)) %>%
  transmute(AOO_category = `Age of Onset Category`,
            FMM_mean = `FMM Mean`,
            FMM_sd = `FMM SD` )

histo_plot <- plot_distributions(histogram_data = case_only_analysis_df,
                               histogram_variable = "VITageonset",
                               dist_df = dist_params,
                               name_col = "AOO_category",
                               mean_col = "FMM_mean",
                               sd_col = "FMM_sd",
                               size = dim(case_only_analysis_df)[1],
                               binwidth = 2)
histo_plot
```

\newpage
## PRS distributions
```{r}
ggplot() +
  geom_density(data = filter(analysis_df, vitiligo == 0), aes(x = no_mhc_classII_prs_scaled), fill = "blue", alpha = 0.7) +
  geom_density(data = filter(analysis_df, vitiligo == 1), aes(x = no_mhc_classII_prs_scaled), fill = "green", alpha = 0.7) +
  labs(title = "Density of No MHC PRS in Vitiligo Cases and Controls",
       x = "No MHC PRS",
       y = "Density") +
  theme_minimal()
```


\newpage
## Associations with Vitiligo Risk (Case-Control)
To verify that the associations look in line with previous estimates,
we check the association of the PRS and the MHC Class II high-risk SNPs
with overall vitiligo risk (i.e. in all cases and controls in GWAS123):

```{r}
#test the association of each PRS with vitiligo risk
scaled_prs_cols <- analysis_df %>%
  select(ends_with("_prs_scaled")) %>%
  names()

phenotypes <- c("vitiligo", "early_onset_vitiligo", "late_onset_vitiligo")
predictors <- c(scaled_prs_cols, "eA00_rs145954018", "generic_rs9271597", "eAOO_hap_carrier")

combos <- crossing(phenotypes, predictors) %>% arrange(desc(phenotypes))

results_df <- pmap_df(combos,
                      ~ logistic_regression(
                        data = analysis_df,
                        covariates = kCovariates,
                        pheno = ..1,
                        prs = ..2)) %>%
  filter(term == PRS) %>%
  select(-term, -estimate, -std_error) %>%
  mutate(pval = formatC(pval, format = "e", digits = 2)) %>%
  mutate(PRS = str_remove(PRS, "_prs_scaled"))  %>%
  rename(`PRS or SNP` = PRS)

pandoc.table(results_df, round=2)
```

These results look generally as I would expect, with the CONFIRMED risk score
having the highest performance with respect to OR per standard deviation of PRS.
The P-values = 0 mean that the P-value is lower than the numerical precision in R,
which I believe is something like P < 1e-300.

Note that the scaling is different for SNPs and for PRS, so the estimates
are not directly comparable (for PRS, the OR is per standard deviation of PRS, 
whereas for SNPs, it is per SNP risk allele).

\newpage
## Associations with Vitiligo Age-of-Onset (Case Only)

I also checked the association with vitiligo age-of-onset
risk (i.e. cases with non-missing age-of-onset in GWAS123):
```{r}
#test the association with vitiligo age-of-onset
aoo_results_df <- pmap_df(list(prs = predictors),
                      ~ aoo_linear_regression(
                        data = case_only_analysis_df,
                        covariates = kCovariates,
                        pheno = "VITageonset",
                        prs = ..1)) %>%
  filter(term == PRS) %>%
  select(-Phenotype, -term, - std_error) %>%
  mutate(pval = formatC(pval, format = "e", digits = 2)) %>%
  mutate(adj_r_sq = formatC(adj_r_sq, digits = 3)) %>%
  mutate(PRS = str_remove(PRS, "_scaled")) %>%
  rename(`PRS or SNP` = PRS)

pandoc.table(aoo_results_df, round=2)
```

The MHC Class II only PRS (P=4.01e-22) is more strongly associated with age-of-onset 
than the confirmed PRS (P=1.59e-15). Nevertheless, there
is modest association between the non-MHC Class II PRS and lower vitiligo
age-of-onset, as we hypothesized (P=7.30e-04).

Note that the scaling is different for SNPs and for PRS, so the estimates
are not directly comparable (for PRS, the estimate is interpreted as years
per standard deviation of PRS, whereas for SNPs, it is interpreted
as years per copy of SNP risk allele).

```{r}
plot1a <- ggplot(analysis_df, aes(y = VITageonset,
                        x = CONFIRMED_prs_scaled)) +
  geom_point(alpha = 0.3, size = 0.8, aes(color = factor(AOO_category))) + 
  geom_smooth(method = "lm", se = TRUE, color="black") +
  labs(title = "A",
       x = "Vitiligo PRS",
       y = "Vitiligo Age-of-Onset",
       color = "Age-of-Onset Category") +
    theme_bw() +
  xlim(-3, 5) +
  geom_text(aes(label = str_c("P-value = ", "1.59e-15")),
             x = 4.9, y = 85, hjust = 1, vjust = 1, size = 3, color = "black") +
  geom_text(aes(label = str_c("\u03B2 = ", "-2.37")),
          x = 4.9, y = 80, hjust = 1, vjust = 1, size = 3, color = "black") +
  theme(legend.position = "none")
  

plot1a
```
```{r include=FALSE}
ggsave(str_c(here(), "/figures/fig1a.svg"), height = 3, width = 6, units = "in")
ggsave(str_c(here(), "/figures/fig1a.tiff"), height = 3, width = 6, units = "in")
```


```{r}
plot1b <- ggplot(analysis_df, aes(y = VITageonset,
                        x = no_mhc_classII_prs_scaled)) +
  geom_point(alpha = 0.3, size = 0.8, aes(color = factor(AOO_category))) + 
  geom_smooth(method = "lm", se = TRUE, color="black") +
  labs(title = "B",
       x = "Vitiligo PRS (MHC Class II Excluded)",
       y = "Vitiligo Age-of-Onset",
       color = "Age-of-Onset Category") +
    scale_color_discrete(labels = c("early_onset" = "Early AOO", 
                     "late_onset" = "Late AOO", 
                     "NA" = "Not Assigned")) +
    theme_bw() +
  xlim(-3, 5) +
  geom_text(aes(label = str_c("P-value = ", "7.3e-04")),
             x = 4.9, y = 85, hjust = 1, vjust = 1, size = 3, color = "black") +
  geom_text(aes(label = str_c("\u03B2 = ", "-1.03")),
          x = 4.9, y = 80, hjust = 1, vjust = 1, size = 3, color = "black") +
  theme(legend.position = "bottom",
        legend.background = element_rect(size = 0.5, linetype = "solid", colour = "black"),
        legend.text = element_text(size = 6),       # Smaller legend text
        legend.title = element_text(size = 6),     # Smaller legend title
        legend.key.size = unit(0.5, "lines"),       # Smaller legend key size
        legend.spacing = unit(0.1, "cm"))           # Smaller spacing between legend items


plot1b
```
```{r include=FALSE}
ggsave(str_c(here(), "/figures/fig1b.svg"), height = 3, width = 6, units = "in")
ggsave(str_c(here(), "/figures/fig1b.tiff"), height = 3, width = 6, units = "in")
```

\newpage
### Stratified Association by High-Risk MHC Haplotype Carrier Status

```{r}
#test the association with vitiligo age-of-onset
case_only_analysis_df <-
  case_only_analysis_df %>%
  mutate(high_risk_MHC_haplotype_carrier =
           ifelse(
             eAOO_hap_carrier==1, "1 or 2 Copies", "0 Copies"
           ))

hap_carrier_res <- aoo_linear_regression(
                        data = filter(case_only_analysis_df, high_risk_MHC_haplotype_carrier=="1 or 2 Copies"),
                        covariates = kCovariates,
                        pheno = "VITageonset",
                        prs = "no_mhc_classII_prs_scaled") %>%
  filter(term == PRS) %>%
  mutate(high_risk_MHC_haplotype_carrier = "1 or 2 Copies")

non_carrier_res <- aoo_linear_regression(
                        data = filter(case_only_analysis_df,high_risk_MHC_haplotype_carrier=="0 Copies"),
                        covariates = kCovariates,
                        pheno = "VITageonset",
                        prs = "no_mhc_classII_prs_scaled") %>%
  filter(term == PRS) %>%
  mutate(high_risk_MHC_haplotype_carrier = "0 Copies")

strat_hap_aoo_results_df <- bind_rows(hap_carrier_res, non_carrier_res) %>%
  select(-term, - std_error) %>%
  mutate(pval = prettyNum(pval, digits = 2)) %>%
   mutate(adj_r_sq = formatC(adj_r_sq, digits = 3)) %>%
  mutate(PRS = str_remove(PRS, "_prs_scaled")) %>%
  rename(`PRS` = PRS) %>%
  select(high_risk_MHC_haplotype_carrier, everything())

pandoc.table(strat_hap_aoo_results_df,
             round=2,
             caption = "Stratified Association between PRS and Age-of-Onset in High Risk Haplotype Carrier Groups")
```

```{r, fig.asp=1.8}
categorized_only_analysis_df <- case_only_analysis_df %>%
  filter(!is.na(high_risk_MHC_haplotype_carrier))

plot1c <- ggplot(categorized_only_analysis_df, aes(y = VITageonset,
                        x = no_mhc_classII_prs_scaled)) +
  geom_point(alpha = 0.3, size = 0.4, aes(color = factor(AOO_category))) + 
  geom_smooth(method = "lm", se = TRUE, color="black") +
  facet_grid(high_risk_MHC_haplotype_carrier ~ .) + 
  labs(title = "C",
       x = "Vitiligo PRS (MHC Class II Excluded)",
       y = "Vitiligo Age-of-Onset",
       color = "Age-of-Onset Category") +
    theme_bw() +
  geom_text(data = strat_hap_aoo_results_df, aes(label = paste0("P-value = ", pval)),
             x = 4.5, y = 85, hjust = 1, vjust = 1, size = 3, color = "black") +
  geom_text(data = strat_hap_aoo_results_df,
          aes(label = paste0("\u03B2 = ", round(estimate, 2))),
          x = 4.5, y = 78, hjust = 1, vjust = 1, size =3, color = "black") +
  guides(color = FALSE) +
  theme(strip.text = element_text(size = 10))


plot1c
```

```{r include=FALSE}
ggsave(str_c(here(), "/figures/fig1c.svg"), height = 6, width = 3, units = "in")
ggsave(str_c(here(), "/figures/fig1c.tiff"), height = 6, width = 3, units = "in")
```

Put all of the Figure 1 panels together...
```{r}
plot1a + plot_spacer() + plot1b  + plot1c + 
  plot_layout(widths = c(2, 1))
```

```{r include=FALSE}
ggsave(str_c(here(), "/figures/fig1_all.svg"), height = 8, width = 6.8, units = "in")
ggsave(str_c(here(), "/figures/fig1_all.tiff"), height = 8, width = 6.8, units = "in")
```

```{r}
aoo_by_hap_summ <- case_only_analysis_df %>%
  group_by(AOO_category, high_risk_MHC_haplotype_carrier) %>%
  summarize(n=n()) %>%
  filter(!is.na(AOO_category) & !is.na(high_risk_MHC_haplotype_carrier)) %>%
  ungroup() %>%
  group_by(AOO_category) %>%
  mutate(sum = sum(n)) %>%
  ungroup() %>%
  mutate(percent = 100 * (n/sum))

pandoc.table(aoo_by_hap_summ)
```



\newpage
### Stratified Association in Early and Late Onset Groups

```{r}
#test the association with vitiligo age-of-onset
eAOO_res <- aoo_linear_regression(
                        data = filter(case_only_analysis_df,
                                      early_onset_vitiligo==1 & high_risk_MHC_haplotype_carrier=="0 Copies"),
                        covariates = kCovariates,
                        pheno = "VITageonset",
                        prs = "no_mhc_classII_prs_scaled") %>%
  filter(term == PRS) %>%
  mutate(AOO_category = "early_onset")

lAOO_res <- aoo_linear_regression(
                        data = filter(case_only_analysis_df,
                                      late_onset_vitiligo==1 & high_risk_MHC_haplotype_carrier=="1 or 2 Copies"),
                        covariates = kCovariates,
                        pheno = "VITageonset",
                        prs = "no_mhc_classII_prs_scaled") %>%
  filter(term == PRS) %>%
  mutate(AOO_category = "late_onset")

strat_aoo_results_df <- bind_rows(eAOO_res, lAOO_res) %>%
  select(-term, - std_error) %>%
  mutate(pval = prettyNum(pval, digits = 2)) %>%
  mutate(adj_r_sq = formatC(adj_r_sq, digits = 3)) %>%
  mutate(PRS = str_remove(PRS, "_prs_scaled")) %>%
  rename(`PRS` = PRS) %>%
  select(AOO_category, everything())

pandoc.table(strat_aoo_results_df,
             round=2,
             caption = "Stratified Association between PRS and Age-of-Onset in Early- and Late-Onset Groups")
```

### Look at assocaition of PRS in MHC High Risk Haplotype Carriers
```{r}
risk_in_hap_carriers <- logistic_regression(data = filter(analysis_df, eAOO_hap_carrier==1),
                        covariates = c(kCovariates),
                        pheno = "vitiligo",
                        prs = "no_mhc_classII_prs_scaled") %>%
  filter(term=="no_mhc_classII_prs_scaled") %>%
  mutate(pval = prettyNum(pval, digits = 2))
  
pandoc.table(risk_in_hap_carriers, round=2)
```



### Compare non-MHC SNP effect estimates in early- and late-onset

Randomly downsample cases so that there are exactly 700 cases with early-onset and
700 cases with late-onset. Also randomly sample two sets of 700 cases each.
Then, compute the association between each of the 
non-MHC lead GWAS variants and disease risk.

```{r}
# Define the function to perform bootstrapping
perform_bootstrap <- function(data,
                              iterations = 100,
                              phenotypes = c("early_onset_vitiligo", "late_onset_vitiligo")) {
  results <- list()
  shuffle_from_col = phenotypes[1]
  phenotypes = c(phenotypes, "shuffled_onset_vitiligo1", "shuffled_onset_vitiligo2")
  
  
  for (i in 1:iterations) {
    # True sample: Sample 700 individuals by AOO_category
    true_sample <- data %>%
      filter(!is.na(AOO_category)) %>%
      group_by(AOO_category) %>%
      sample_n(700, replace = TRUE) %>%
      ungroup() %>%
      #null sample: sample two sets of vitiligo cases and randomly assign AOO category
      mutate(shuffled_onset_vitiligo1 = sample(x=c(NA,1), prob=c(0.5, 0.5), size=1400, replace=TRUE)) %>%
      mutate(shuffled_onset_vitiligo2 = sample(x=c(NA,1), prob=c(0.5, 0.5), size=1400, replace=TRUE)) %>%
      bind_rows(filter(data, vitiligo == 0)) %>%
      mutate(shuffled_onset_vitiligo1 = ifelse(vitiligo==0, 0, shuffled_onset_vitiligo1)) %>%
      mutate(shuffled_onset_vitiligo2 = ifelse(vitiligo==0, 0, shuffled_onset_vitiligo2))
    
    #setup to run logit across all snps
    predictors <- data %>%
      select(starts_with("RS")) %>%
      colnames(.)
    
    combos <- crossing(phenotypes, predictors) %>%
      arrange(desc(phenotypes))
    
    res_df <- pmap_df(combos,
                      ~ logistic_regression(
                        data = true_sample,
                        covariates = kCovariates,
                        pheno = ..1,
                        prs = ..2)) %>%
      filter(term == PRS) %>%
      rename(SNP = PRS) %>%
      transmute(iteration = i,
                Phenotype,
                SNP,
                estimate) %>%
      pivot_wider(names_from = c(Phenotype), values_from = c(estimate)) %>%
      mutate(early_v_late_estimate=early_onset_vitiligo - late_onset_vitiligo) %>%
      mutate(shuffled_v_shuffled_estimate=shuffled_onset_vitiligo1 - shuffled_onset_vitiligo2)
    
    results[[i]] <- res_df
  }
  combined_results <- bind_rows(results)
  return(combined_results)
}

filepath <- str_c(here(), "/bootstrap_compare_early_v_late_100_iterations.csv")

if (file.exists(filepath)) {
  # If the file exists, read the data
  bootstrap_results <- fread(filepath)
} else {
  # If the file does not exist, run the bootstrap function and save the results
  bootstrap_results <- perform_bootstrap(analysis_df, iterations = 100)
  write.csv(bootstrap_results, filepath, quote=FALSE, row.names = FALSE)
}
```



```{r}
# Use the bootstrap results to determine whether there is a true difference
# in effect estimates

snp_estimate_res = list()
for(i in unique(bootstrap_results$iteration)){
  tmp_dat <- bootstrap_results %>% filter(iteration==i)
  
  #compare whether a zero-intercept model yields a slope < 1 for early vs late onset
  formula <- as.formula(paste("early_onset_vitiligo", "~", "late_onset_vitiligo -1"))
  compare_snps_mod <- lm(formula,
             data = filter(tmp_dat,
              !SNP %in% c("RS145954018", "RS114448410")))
  hypothesis <- c("late_onset_vitiligo = 1")
  test_compare_snps_mod <- linearHypothesis(compare_snps_mod, hypothesis)
  pval_compare_snps <- test_compare_snps_mod$`Pr(>F)`[2]
  slope_compare_snps <- compare_snps_mod$coefficients[1]
  
  
  #compare whether a zero-intercept model yields a slope < 1 for random case subsets
    null_formula <- as.formula(paste("shuffled_onset_vitiligo1", "~", "shuffled_onset_vitiligo2 -1"))
    null_compare_snps_mod <- lm(null_formula,
             data = filter(tmp_dat,
              !SNP %in% c("RS145954018", "RS114448410")))
    null_hypothesis <- c("shuffled_onset_vitiligo2 = 1")
    null_test_compare_snps_mod <- linearHypothesis(null_compare_snps_mod, null_hypothesis)
    null_pval_compare_snps <- null_test_compare_snps_mod$`Pr(>F)`[2]
    null_slope_compare_snps <- null_compare_snps_mod$coefficients[1]
  
  
  #put the results together into a table
  tmp_res <- tribble(
    ~"iteration", ~"slope", ~"pvalue", ~"null_slope", ~"null_pvalue",
    i, slope_compare_snps, pval_compare_snps, null_slope_compare_snps, null_pval_compare_snps 
  )
  snp_estimate_res[[i]] <-  tmp_res
}
snp_estimate_res <- bind_rows(snp_estimate_res)

#perform t-test on bootstrap results (early v late slope vs. random v. random slope)
test_result <- t.test(snp_estimate_res$null_slope, snp_estimate_res$slope) %>% tidy()
slope = test_result %>% pull(estimate2)
pval = test_result %>% pull(p.value)
pandoc.table(test_result)
```
```{r}
#Get 95% CI for estimate #2
slope_ci <- t.test(snp_estimate_res$slope, mu = 1) %>% tidy()
pandoc.table(slope_ci)
```


```{r}
#use the bootstrap results results to get mean effect est
early_v_late_snp_estimate <- bootstrap_results %>%
  group_by(SNP) %>%
  summarise(
    mean_early_onset_vitiligo = mean(early_onset_vitiligo),
    sd_early_onset_vitiligo = sd(early_onset_vitiligo),
    mean_late_onset_vitiligo = mean(late_onset_vitiligo),
    sd_late_onset_vitiligo = sd(late_onset_vitiligo),
    mean_early_v_late_estimate = mean(early_v_late_estimate),
    sd_early_v_late_estimate = sd(early_v_late_estimate)
  ) %>%
  mutate(predicted_early_onset = slope*mean_late_onset_vitiligo)
early_v_late_snp_estimate
```



Below, I have plotted the effect estimates in the early-onset and late-onset group together. If the other loci have generally equivalent effects in both groups, we expect that the points should generally fall around the y=x line (denoted in red).  If, on the other hand, there is effect size dilution in the early-onset group, we expect that the slope will be <1.

```{r, fig.align='center'}

##make the plot
plot2 <- ggplot(filter(early_v_late_snp_estimate,
              !SNP %in% c("RS145954018", "RS114448410")),
       aes(x = mean_late_onset_vitiligo,
           y = mean_early_onset_vitiligo)) +
  geom_point(color="grey") +
  geom_errorbar(aes(
    ymin = mean_early_onset_vitiligo - sd_early_onset_vitiligo,
    ymax = mean_early_onset_vitiligo + sd_early_onset_vitiligo),
    width = 0.02,
    alpha = 0.2) +
    geom_errorbarh(aes(
    xmin = mean_late_onset_vitiligo - sd_late_onset_vitiligo,
    xmax = mean_late_onset_vitiligo + sd_late_onset_vitiligo),
    height = 0.02, 
    alpha = 0.2) +
  theme_classic() +
  geom_abline(intercept=0, slope=1, color="black", linetype="dashed") +
  geom_abline(intercept=0, slope=slope, color = "blue") +
  xlim(-0.02,0.73) +
  ylim(-0.02,0.73) +
  xlab(paste0("Late-Onset Vitiligo SNP ", "\u03B2", "s")) +
  ylab(paste0("Early-Onset Vitiligo SNP ", "\u03B2", "s")) +
  geom_text(aes(label = paste0("Bootstrap P-value = 6.75e-15")),
             x = 0.39, y = 0.7, hjust = 0, vjust = 0, size = 3, color = "black") +
  geom_text(aes(label = paste0("Mean Bootstrap \u03B2 = ", round(slope, 3))),
          x = 0.39, y = 0.67, hjust = 0, vjust = 0, size = 3, color = "black")


plot2
```

```{r include=FALSE}
ggsave(str_c(here(), "/figures/fig2.svg"), height = 5, width = 5, units = "in")
ggsave(str_c(here(), "/figures/fig2.tiff"), height = 5, width = 5, units = "in")
```


